### PR TITLE
Issue83 and Issue118

### LINKED ISSUE #118: Display first name and first letter of last name in leaderboard
On the leaderboard, rather than display the login name (UTORID), we display a variation of their real name. For example, "Larry Z", "Ahmed Q", etc. This feels more personal and less uncomfortable for some people than displaying the full name like "Larry Zhang".

### PR DESCRIPTION
**Pull request by Mohamed-A-Mohamed**
This PR fiixes #83 and fixes #118. 

**The images for the podium and the ones in the table are temporary until users have profile pictures.**

peek 2017-11-24 22-57

### CODE REVIEW

File: public/css/leaderboard-table.css [hunk 1]
```diff
@@ -29,3 +29,9 @@
 .col-isuser {
   font-weight: bold;
 }
+
+img.responsive-img, video.responsive-video {
```
Comment (ahmedqarmout2) [inline public/css/leaderboard-table.css]: what does this do? also this css and the next css file have different indentation, please fix

File: public/css/leaderboard.css [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+li.gold .score {
+    background-color: rgba(255,215,0,0.7);
+}
+
+li.silver .score {
+    background-color: rgba(192,192,192,0.7);
+}
+
+li.bronze .score {
```
Comment (ahmedqarmout2) [inline public/css/leaderboard.css:57]: can we use more proper pictures please

File: public/js/home.js [hunk 3]
```diff
@@ -79,13 +79,14 @@ var sortRequest = function(type) {
 /*
  * Fetch the mini leaderboard table and display it in the sidebar.
  */
-var fetchLeaderboard = function() {
+var fetchLeaderboard = function(t) {
```
Comment (ahmedqarmout2) [inline public/js/home.js]: use something more descriptive than t

File: public/js/home.js [hunk 4]
```diff
@@ -98,4 +99,4 @@ var fetchLeaderboard = function() {
     });
 }
 
-fetchLeaderboard();
+fetchLeaderboard('overall');
```
Comment (ahmedqarmout2) [inline public/js/home.js]: put this in the common file in public/js

File: public/js/leaderboard.js [hunk 5]
```diff
@@ -2,18 +2,51 @@
 $('#nav-leaderboard').addClass('active');
 
 /* Get HTML for the complete leaderboard table from the server and display it. */
-var fetchLeaderboard = function() {
+var fetchLeaderboard = function(t) {
+
```
Comment (ahmedqarmout2) [inline public/js/leaderboard.js]: remove this space, and use proper variable names

File: public/js/leaderboard.js [hunk 6]
```diff
@@ -2,18 +2,51 @@
...
+            type: t
         },
         success: function(data) {
             $('.leaderboard-content').html(data);
         }
     });
 }
 
-fetchLeaderboard();
+fetchLeaderboard('overall');
```
Comment (ahmedqarmout2) [inline public/js/leaderboard.js]: if you are only using overall, then why have it?

File: server.js [hunk 7]
```diff
@@ -196,6 +196,9 @@ const chooseAllForm = pug.compileFile('views/question_types/chooseAll-answer.pug
 const matchingForm = pug.compileFile('views/question_types/matching-answer.pug');
 const orderingForm = pug.compileFile('views/question_types/ordering-answer.pug');
 const leaderboardTable = pug.compileFile('views/leaderboard-table.pug');
+const pointsTable = pug.compileFile('views/points-leaderboard.pug');
```
Comment (ahmedqarmout2) [inline server.js]: move all of these inside a folder called leaderboard (you will have to make this folder)

Comment (ahmedqarmout2) [reply server.js]: all of these === any pug file related to leaderboard

File: public/js/leaderboard.js [hunk 8]
```diff
@@ -2,18 +2,51 @@
...
+
```
Comment (ahmedqarmout2) [inline public/js/leaderboard.js]: remove the space here

File: public/js/leaderboard.js [hunk 9]
```diff
@@ -2,18 +2,51 @@
...
+$('#option-overall').click(function(evt) {
+    $('#points-board-card').addClass('hidden');
+    $('#accuracy-board-card').addClass('hidden');
+    $('#overall-board-card').removeClass('hidden');
```
Comment (ahmedqarmout2) [inline public/js/leaderboard.js]: what about '#attempt-board-card' dont you have to hide it

Comment (ahmedqarmout2) [reply public/js/leaderboard.js]: same goes below

File: server.js [hunk 10]
```diff
@@ -209,14 +212,77 @@ app.get('/leaderboard-table', function(req, res) {
 
     ft = true;
     shrt = false;
+    
+    if (req.query.fullTable == 'false')
```
Comment (ahmedqarmout2) [inline server.js]: why are you adding those, are we actually ever sending fullTable  as a different value

Comment (ahmedqarmout2) [reply server.js]: always use ===

Comment (ahmedqarmout2) [reply server.js]: if (...) {
}
always use brackets

File: server.js [hunk 11]
```diff
@@ -209,14 +212,77 @@ app.get('/leaderboard-table', function(req, res) {
...
+        ft = false;
+    if (req.query.longTable == 'false')
+        shrt = true;
+    
+    if(req.query.type === 'points'){
```
Comment (ahmedqarmout2) [inline server.js]: use a try catch and set the default to return 500 with an error message

File: public/css/leaderboard.css [hunk 12]
```diff
@@ -0,0 +1,61 @@
+
```
Comment (esteeroramy) [inline public/css/leaderboard.css]: extra line

File: server.js [hunk 13]
```diff
@@ -209,14 +212,77 @@ app.get('/leaderboard-table', function(req, res) {
...
+                userid: req.session.user.id
+            });
+    
+            return res.status(200).send(html);
+        });
+    }
+});
 
+/* Fetch and render the points table. Send HTML as response. */
+app.get('/points-leaderboard', function(req, res) {
```
Comment (ahmedqarmout2) [inline server.js]: why does this have a different path? I thought you are only sorting in the frontend. Also isnt this covered above

File: public/css/leaderboard.css (hunk 2 above)
Comment (esteeroramy) [reply public/css/leaderboard.css:57]: can we use the logo until profile pictures are done?

File: server.js [hunk 14]
```diff
@@ -209,14 +212,77 @@ app.get('/leaderboard-table', function(req, res) {
...
+    shrt = false;
+    
     if (req.query.fullTable == 'false')
         ft = false;
     if (req.query.longTable == 'false')
         shrt = true;
 
     lb.leaderboard(req.session.user.id, shrt, function(leader) {
-        var html = leaderboardTable({
+        var html = pointsTable({
```
Comment (ahmedqarmout2) [inline server.js]: misleading name. We dont have a points table we have a leaderboard table thats sorted by points, why do we even have a different table for this

File: public/js/leaderboard.js [hunk 15]
```diff
@@ -2,18 +2,51 @@
 $('#nav-leaderboard').addClass('active');
```
Comment (esteeroramy) [inline public/js/leaderboard.js:2]: why cant we add this in the pug directly

File: public/js/leaderboard.js [hunk 16]
```diff
@@ -2,18 +2,51 @@
...
+
+$('#option-overall').click(function(evt) {
+    $('#points-board-card').addClass('hidden');
+    $('#accuracy-board-card').addClass('hidden');
+    $('#overall-board-card').removeClass('hidden');
+    fetchLeaderboard('overall');
+});
+
+$('#option-points').click(function(evt) {
+    $('#overall-board-card').addClass('hidden');
```
Comment (esteeroramy) [inline public/js/leaderboard.js]: this section of code seems to be repeated a lot, cant we group this in a function?

File: public/js/leaderboard.js [hunk 17]
```diff
@@ -2,18 +2,51 @@
...
+});
+
+$('#option-attempt').click(function(evt) {
+    $('#overall-board-card').addClass('hidden');
+    $('#points-board-card').addClass('hidden');
+    $('#accuracy-board-card').addClass('hidden');
+    $('#attempt-board-card').removeClass('hidden');    
+    fetchLeaderboard('attempt');
+});
+
```
Comment (esteeroramy) [inline public/js/leaderboard.js]: extra lines down here

File: views/accuracy-leaderboard.pug [hunk 18]
```diff
@@ -0,0 +1,39 @@
+.row#accuracy-board-card
```
Comment (ahmedqarmout2) [inline views/accuracy-leaderboard.pug]: use the same naming system, leaderboard-<type>

File: views/accuracy-leaderboard.pug [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+  .col.s12
+    table.bordered.centered
+      thead
+        tr
+          th Rank
+          th User
+          th Accuracy
+
```
Comment (ahmedqarmout2) [inline views/accuracy-leaderboard.pug]: remove space

File: server.js [hunk 20]
```diff
@@ -209,14 +212,77 @@ app.get('/leaderboard-table', function(req, res) {
 
     ft = true;
```
Comment (esteeroramy) [inline server.js]: can we use more descriptive names, not really sure what these variables are doing

File: views/accuracy-leaderboard.pug [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+            
+      tbody
+        each s in leaderboard
+          if s.rank == 0
+            tr.col-default
+              td ...
+              td ...
+              td ...
+          else
+            if s.points != 0 && s.rank <= 3
```
Comment (ahmedqarmout2) [inline views/accuracy-leaderboard.pug]: !==

File: views/accuracy-leaderboard.pug [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+              if s.rank == 1
```
Comment (ahmedqarmout2) [inline views/accuracy-leaderboard.pug]: ===

Comment (ahmedqarmout2) [reply views/accuracy-leaderboard.pug]: for the entire file

File: views/accuracy-leaderboard.pug [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+                - rowclass = 'col-bronze';
+            else
+              - rowclass = 'col-default';
+            if s.id == userid
+              - rowclass += ' col-isuser';
+            tr(class=rowclass)
+              td #{s.rank}
+              td
+                img.responsive-img.circle(src='img/leaderboard_profile.png')
+                | #{s.fname + ' ' + s.lname.substring(0, 1) + '.'}
```
Comment (ahmedqarmout2) [inline views/accuracy-leaderboard.pug]: make a function for this, dont do the computation in pug

File: server.js [hunk 24]
```diff
@@ -209,14 +212,77 @@ app.get('/leaderboard-table', function(req, res) {
...
+    }
+});
 
+/* Fetch and render the points table. Send HTML as response. */
+app.get('/points-leaderboard', function(req, res) {
+    if (!req.session.user) {
+        return res.redirect('/');
+    }
+
+    var ft, shrt;
```
Comment (esteeroramy) [inline server.js]: why do you need to define these variables then assign them? you can just do:
``` javascript
var variableName = true;
```

File: server/leaderboard.js [hunk 25]
```diff
@@ -27,23 +27,125 @@ var users = require('./users.js');
...
+exports.leaderboard = function(userid, shrt, type, callback) {
     users.getStudentsListSorted(0, function(err, studentlist) {
+
         var rank = 0;
         var last = -1;
         var userind;
 
+        //Sort the list of students depending on the type (defualt sorting is by points)
+        if (type === 'accuracy'){
+            studentlist = studentlist.sort(function(a,b) {
```
Comment (esteeroramy) [inline server/leaderboard.js]: would be a good idea to use names other than a,b,x, and y

File: server/leaderboard.js [hunk 26]
```diff
@@ -27,23 +27,125 @@ var users = require('./users.js');
...
  */
-exports.leaderboard = function(userid, shrt, callback) {
+exports.leaderboard = function(userid, shrt, type, callback) {
     users.getStudentsListSorted(0, function(err, studentlist) {
+
         var rank = 0;
         var last = -1;
         var userind;
 
+        //Sort the list of students depending on the type (defualt sorting is by points)
```
Comment (esteeroramy) [inline server/leaderboard.js]: space after `//` for consistency

File: views/leaderboard-table.pug [hunk 27]
```diff
@@ -1,44 +1,41 @@
...
-        tr(class=rowclass)
-          td #{s.rank}
-          td #{s.id}
-          td #{s.points}
-          if fullTable
-            td #{s.totalAttemptsCount}
-            td #{s.correctAttemptsCount}
-            if s.totalAttemptsCount === 0
-              td 0%
+.row#overall-board-card
```
Comment (ahmedqarmout2) [inline views/leaderboard-table.pug]: whats the sorting type of this leaderboard?

File: server/leaderboard.js [hunk 28]
```diff
@@ -27,23 +27,125 @@ var users = require('./users.js');
  *
  * If shrt is true, return leaderboard with max eight entries.
  */
-exports.leaderboard = function(userid, shrt, callback) {
+exports.leaderboard = function(userid, shrt, type, callback) {
     users.getStudentsListSorted(0, function(err, studentlist) {
+
```
Comment (ahmedqarmout2) [inline server/leaderboard.js]: remove space

File: server/leaderboard.js [hunk 29]
```diff
@@ -27,23 +27,125 @@ var users = require('./users.js');
...
         var last = -1;
         var userind;
 
+        //Sort the list of students depending on the type (defualt sorting is by points)
+        if (type === 'accuracy'){
+            studentlist = studentlist.sort(function(a,b) {
+                x = a.correctAttemptsCount/a.totalAttemptsCount;
+                y = b.correctAttemptsCount/b.totalAttemptsCount;
+
+                if (a.totalAttemptsCount === 0 && b.totalAttemptsCount ===0){
```
Comment (esteeroramy) [inline server/leaderboard.js]: seems like there is a good amount of duplicated code here, this should be moved in a helper function

File: server/leaderboard.js [hunk 30]
```diff
@@ -27,23 +27,125 @@ var users = require('./users.js');
...
-exports.leaderboard = function(userid, shrt, callback) {
+exports.leaderboard = function(userid, shrt, type, callback) {
     users.getStudentsListSorted(0, function(err, studentlist) {
+
         var rank = 0;
         var last = -1;
         var userind;
 
+        //Sort the list of students depending on the type (defualt sorting is by points)
+        if (type === 'accuracy'){
```
Comment (ahmedqarmout2) [inline server/leaderboard.js]: if (...) { <==== the space

File: server/leaderboard.js [hunk 31]
```diff
@@ -27,23 +27,125 @@ var users = require('./users.js');
...
+                    return -1;
+                } else if (a.totalAttemptsCount === 0){
+                    return 1;
+                } else{
+                    return y - x;
+                }
+            })
+        } else if (type === 'overall'){
+            studentlist = studentlist.sort(function(a,b) {
+                x = a.points * ((a.correctAttemptsCount / a.totalAttemptsCount) + (a.points / a.totalAttemptsCount));
```
Comment (esteeroramy) [inline server/leaderboard.js]: lots of math happening here, please add a comment to explain it

File: server/leaderboard.js (hunk 30 above)
Comment (ahmedqarmout2) [reply server/leaderboard.js]: use a switch and return an error message at the end

File: server/leaderboard.js [hunk 32]
```diff
@@ -27,23 +27,125 @@ var users = require('./users.js');
...
-            if (studentlist[i].points != last) {
-                rank = i + 1;
-                last = studentlist[i].points;
-            }
-            studentlist[i].rank = rank;
+            
+            //Rank is assigned depending on the type
+
+            if(type === 'points'){
+                if (studentlist[i].points != last) {
```
Comment (esteeroramy) [inline server/leaderboard.js]: `!==`

File: server/leaderboard.js [hunk 33]
```diff
@@ -27,23 +27,125 @@ var users = require('./users.js');
...
+                    rank = i + 1;
+                    last = studentlist[i].points;
+                }
+                studentlist[i].rank = rank;
+    
+                if (studentlist[i].id == userid)
+                    userind = i;
+
+            } else if(type === 'attempt'){
```
Comment (esteeroramy) [inline server/leaderboard.js]: these types should probably be moved to `common.js`

File: server/leaderboard.js [hunk 34]
```diff
@@ -27,23 +27,125 @@ var users = require('./users.js');
...
+            //Rank is assigned depending on the type
+
+            if(type === 'points'){
+                if (studentlist[i].points != last) {
+                    rank = i + 1;
+                    last = studentlist[i].points;
+                }
+                studentlist[i].rank = rank;
+    
+                if (studentlist[i].id == userid)
```
Comment (esteeroramy) [inline server/leaderboard.js]: `===`

File: server/leaderboard.js [hunk 35]
```diff
@@ -27,23 +27,125 @@ var users = require('./users.js');
...
+                y = b.correctAttemptsCount/b.totalAttemptsCount;
+
+                if (a.totalAttemptsCount === 0 && b.totalAttemptsCount ===0){
+                    return 0;
+                } else if (b.totalAttemptsCount === 0){
+                    return -1;
+                } else if (a.totalAttemptsCount === 0){
+                    return 1;
+                } else{
+                    return y - x;
```
Comment (ahmedqarmout2) [inline server/leaderboard.js]: we use a return when there is a callback?

File: server/leaderboard.js [hunk 36]
```diff
@@ -27,23 +27,125 @@ var users = require('./users.js');
  *
  * If shrt is true, return leaderboard with max eight entries.
  */
-exports.leaderboard = function(userid, shrt, callback) {
+exports.leaderboard = function(userid, shrt, type, callback) {
```
Comment (ahmedqarmout2) [inline server/leaderboard.js]: not type, use sort type

File: server/leaderboard.js [hunk 37]
```diff
@@ -27,23 +27,125 @@ var users = require('./users.js');
...
+                }
+                studentlist[i].rank = rank;
+    
+                if (studentlist[i].id == userid)
+                    userind = i;
+
+            } else if(type === 'overall'){
+                var overall = 0;
+                if (studentlist[i].totalAttemptsCount != 0){
+                    overall = studentlist[i].points * ((studentlist[i].correctAttemptsCount / studentlist[i].totalAttemptsCount) + (studentlist[i].points / studentlist[i].totalAttemptsCount));
```
Comment (esteeroramy) [inline server/leaderboard.js]: very confusing, this should be broken up to individual pieces

File: server/leaderboard.js [hunk 38]
```diff
@@ -27,23 +27,125 @@ var users = require('./users.js');
...
         for (var i = 0; i < studentlist.length; ++i) {
             /* subsequent users with same amount of points have same rank */
-            if (studentlist[i].points != last) {
-                rank = i + 1;
-                last = studentlist[i].points;
-            }
-            studentlist[i].rank = rank;
+            
+            //Rank is assigned depending on the type
+
```
Comment (ahmedqarmout2) [inline server/leaderboard.js]: remove space

File: views/accuracy-leaderboard.pug [hunk 39]
```diff
@@ -0,0 +1,39 @@
...
+        tr
+          th Rank
+          th User
+          th Accuracy
+            
+      tbody
+        each s in leaderboard
+          if s.rank == 0
+            tr.col-default
+              td ...
```
Comment (esteeroramy) [inline views/accuracy-leaderboard.pug]: we can use a loop here

File: views/accuracy-leaderboard.pug [hunk 40]
```diff
@@ -0,0 +1,39 @@
...
+                - rowclass = 'col-gold';
+              if s.rank == 2
+                - rowclass = 'col-silver';
+              if s.rank == 3
+                - rowclass = 'col-bronze';
+            else
+              - rowclass = 'col-default';
+            if s.id == userid
+              - rowclass += ' col-isuser';
+            tr(class=rowclass)
```
Comment (esteeroramy) [inline views/accuracy-leaderboard.pug]: can't you just do
``` pug
tr.rowclass
```

File: views/leaderboard-table.pug [hunk 41]
```diff
@@ -1,44 +1,41 @@
...
+          th Overall Score
+
+      tbody
+        each s in leaderboard
+          if s.rank == 0
+            tr.col-default
+              td ...
+              td ...
+              td ...
+
```
Comment (ahmedqarmout2) [inline views/leaderboard-table.pug]: remove space

File: views/accuracy-leaderboard.pug [hunk 42]
```diff
@@ -0,0 +1,39 @@
...
+              td #{s.rank}
+              td
+                img.responsive-img.circle(src='img/leaderboard_profile.png')
+                | #{s.fname + ' ' + s.lname.substring(0, 1) + '.'}
+              if s.totalAttemptsCount === 0
+                td 0.00%
+              else
+                td #{((s.correctAttemptsCount / s.totalAttemptsCount) * 100).toFixed(2)}%
```
Comment (esteeroramy) [inline views/accuracy-leaderboard.pug]: might not be a good idea to have math here, do you think this should be moved somewhere else

File: views/leaderboard-table.pug [hunk 43]
```diff
@@ -1,44 +1,41 @@
...
+        tr
+          th Rank
+          th User
+          th Overall Score
+
+      tbody
+        each s in leaderboard
+          if s.rank == 0
+            tr.col-default
+              td ...
```
Comment (esteeroramy) [inline views/leaderboard-table.pug]: loop

File: views/points-leaderboard.pug [hunk 44]
```diff
@@ -0,0 +1,36 @@
+.row#points-board-card
```
Comment (esteeroramy) [inline views/points-leaderboard.pug]: these leader board files seem to be very much like each other, the functionality should probably be extracted into a helper component and used through `include` functionality

**Review by esteeroramy — CHANGES_REQUESTED** [review]

File: views/leaderboard.pug [hunk 45]
```diff
@@ -15,11 +15,37 @@ block content
   .container
     .row
       .col.s12
-        .card.center
-          .card-content
-            h4.center Leaderboard
-          .card-action
-            .leaderboard-content
+        .row
```
Comment (ahmedqarmout2) [inline views/leaderboard.pug]: what is this leaderboard for?

File: views/leaderboard.pug [hunk 46]
```diff
@@ -15,11 +15,37 @@ block content
...
+          ul.tabs
+            li.tab.col.s3
+              a.noLineLink.active#option-overall(href='#') Overall
```
Comment (ahmedqarmout2) [inline views/leaderboard.pug]: whats the href for?

**Comment by esteeroramy** [comment]
Added some comments, please fix it up and I will take a look at it again to see things that I might have missed

File: views/leaderboard.pug [hunk 47]
```diff
@@ -15,11 +15,37 @@ block content
...
+          .leaderboard-content
+      .col.s6
+        .card.white
+          .podium
+            p
+              | Top Score : 
+              span.score [Highest Score Will Go Here]
+            ol
+              li.gold
+                img(alt='first', src='img/leaderboard_profile.png')
```
Comment (ahmedqarmout2) [inline views/leaderboard.pug]: lets use proper images

**Comment by esteeroramy** [comment]
Also, can we have that `top 3` card at the top of the page, and expand the leader board horizontally?

File: public/css/leaderboard-table.css (hunk 1 above)
Comment (Mohamed-A-Mohamed) [reply public/css/leaderboard-table.css]: This is css of the pictures in the table beside the students name. I used in the responsive class but had to tweek it.

File: public/css/leaderboard.css (hunk 2 above)
Comment (Mohamed-A-Mohamed) [reply public/css/leaderboard.css:57]: I'll change it to the logo :cry:

File: public/js/leaderboard.js (hunk 9 above)
Comment (Mohamed-A-Mohamed) [reply public/js/leaderboard.js]: Good catch, looks like I missed it.

File: server.js (hunk 10 above)
Comment (Mohamed-A-Mohamed) [reply server.js]: Good catch again, I'll remove fullTable

File: server.js (hunk 13 above)
Comment (Mohamed-A-Mohamed) [reply server.js]: Woops, this is from earlier on and I forgot to remove it.

File: public/css/leaderboard.css (hunk 2 above)
Comment (ahmedqarmout2) [reply public/css/leaderboard.css:57]: why are we using profile pictures here?

File: server.js (hunk 14 above)
Comment (Mohamed-A-Mohamed) [reply server.js]: Good point I'll change the name (I'll also do the same for the others as they're also called table). As for why we have a diffrent table for points, Larry wanted a points table :sweat_smile:

File: views/leaderboard-table.pug (hunk 27 above)
Comment (Mohamed-A-Mohamed) [reply views/leaderboard-table.pug]: It's sorted by the oveall score, overall score = total points * (accuracy + points per attempt / 100)

File: server/leaderboard.js (hunk 35 above)
Comment (Mohamed-A-Mohamed) [reply server/leaderboard.js]: Good catch

File: public/js/leaderboard.js (hunk 15 above)
Comment (Mohamed-A-Mohamed) [reply public/js/leaderboard.js:2]: I didn't add this so not sure :laughing:

File: public/js/leaderboard.js (hunk 16 above)
Comment (Mohamed-A-Mohamed) [reply public/js/leaderboard.js]: Good point, I'll try to goup them.

File: server.js (hunk 20 above)
Comment (Mohamed-A-Mohamed) [reply server.js]: This will be removed as we don't use full table anymore, as pointed out by Ahmed.

File: server.js (hunk 24 above)
Comment (Mohamed-A-Mohamed) [reply server.js]: This was there from before, I'll change it though.

File: server/leaderboard.js (hunk 25 above)
Comment (Mohamed-A-Mohamed) [reply server/leaderboard.js]: I meant to chagne them but I forgot :sweat_smile:

File: server/leaderboard.js (hunk 29 above)
Comment (Mohamed-A-Mohamed) [reply server/leaderboard.js]: Good point

File: views/accuracy-leaderboard.pug (hunk 39 above)
Comment (Mohamed-A-Mohamed) [reply views/accuracy-leaderboard.pug]: It's already in a loop?

File: server/leaderboard.js (hunk 37 above)
Comment (Mohamed-A-Mohamed) [reply server/leaderboard.js]: I'll break it up.

File: views/accuracy-leaderboard.pug (hunk 40 above)
Comment (Mohamed-A-Mohamed) [reply views/accuracy-leaderboard.pug]: Good catch, I'll change it

File: views/leaderboard-table.pug (hunk 43 above)
Comment (Mohamed-A-Mohamed) [reply views/leaderboard-table.pug]: already in a loop?

File: views/points-leaderboard.pug (hunk 44 above)
Comment (Mohamed-A-Mohamed) [reply views/points-leaderboard.pug]: I'll clean them up

File: views/leaderboard.pug (hunk 46 above)
Comment (Mohamed-A-Mohamed) [reply views/leaderboard.pug]: Woops, another thing from earlier on that I forogt to remove.

File: views/leaderboard.pug (hunk 45 above)
Comment (Mohamed-A-Mohamed) [reply views/leaderboard.pug]: This gets filled with what ever fetchLeaderboard returns.

File: views/accuracy-leaderboard.pug (hunk 42 above)
Comment (Mohamed-A-Mohamed) [reply views/accuracy-leaderboard.pug]: Accuracy (same goes for the other calculations) aren't stored anywhere which is why I calculate them here. I can try to pass them through from when I'm assigning the rank.

**Review by mehtadarshan1 — COMMENTED** [review]
Nice stuff, just one question, 
Did you implement the user profile photo upload? Or is this profile picture going to be the same for every one?

**Comment by esteeroramy** [comment]
Any updates on this? Also seems like you have a few conflicts, please make sure that you are on latest develop.

File: public/js/common.js [hunk 48]
```diff
@@ -25,6 +25,13 @@ const snack = Object.freeze({
 
 const loadingAnimation = '<div class="progress"><div class="indeterminate"></div></div>';
 
+const leaderboardTypes = Object.freeze({
+    OVERALL  : {name: 'overall', displayName: 'Overall'},
+    POINTS   : {name: 'points', displayName: 'Points'},
+    ACCURACY : {name: 'accuracy', displayName: 'Accuracy'},
+    ATTEMPT  : {name: 'attempt', displayName: 'Points Per Attempt'}
```
Comment (ahmedqarmout2) [inline public/js/common.js]: Use a different name plz, this name is different than what its used for

File: public/js/home.js [hunk 49]
```diff
@@ -89,18 +92,30 @@ var fetchLeaderboard = function() {
...
-            if (data['status'] === 401) {
-                window.location.href = '/';
-            }
+            leaderboardTable = $(data.leaderboardTableHTML);
+            leaderboardRow = $(data.leaderboardRowHTML);
+            studentLeaderList = data.leader;
+            $('.leaderboard-small').html(leaderboardTable);
+            $('#userRank').html('Current Rank: ' + data.displayRank);            
+            displayLeaderboard(studentLeaderList, data.userId);
         }
```
Comment (ahmedqarmout2) [inline public/js/home.js]: Catch the fail plz

File: public/js/leaderboard.js [hunk 50]
```diff
@@ -1,19 +1,99 @@
...
+            smallBoard: false
         },
         success: function(data) {
-            $('.leaderboard-content').html(data);
+            setPodiumImages();
+            leaderboardTable = $(data.leaderboardTableHTML);
+            leaderboardRow = $(data.leaderboardRowHTML);
+            studentLeaderList = data.leader;
+            $('.leaderboard-content').html(leaderboardTable);
+            displayNewLeaderboard(boardType);
```
Comment (ahmedqarmout2) [inline public/js/leaderboard.js]: Catch the ajax fail

File: server.js [hunk 51]
```diff
@@ -38,6 +38,23 @@ const app = express();
 const port = process.env.QUIZZARD_PORT || 8000;
 
 /* Pre-compiled Pug views */
+const studentTable = pug.compileFile('views/account-table.pug');
```
Comment (ahmedqarmout2) [inline server.js]: Why are these back?

File: server.js [hunk 52]
```diff
@@ -213,32 +229,27 @@ app.get('/about', function(req,res) {
...
+        smallBoard = true;
     }
-
-    users.getLeaderboard(req.session.user._id, shortTable, function(leader) {
-        var html = leaderboardTablePug({
-            fullTable: fullTable,
-            shortTable: shortTable,
-            leaderboard: leader,
-            userid: req.session.user._id
+    users.getLeaderboard(req.session.user._id, smallBoard, function(leader, displayRank) {
```
Comment (ahmedqarmout2) [inline server.js]: Check for err, also what does leader mean?

File: public/js/leaderboard.js [hunk 53]
```diff
@@ -1,19 +1,99 @@
...
         type: 'GET',
         url: '/leaderboard-table',
         data: {
-            fullTable: true,
-            longTable: true
+            smallBoard: false
         },
         success: function(data) {
-            $('.leaderboard-content').html(data);
+            setPodiumImages();
```
Comment (esteeroramy) [inline public/js/leaderboard.js]: this should happen when we are setting up the table

File: public/js/leaderboard.js [hunk 54]
```diff
@@ -1,19 +1,99 @@
...
+            } else {
+                rank = index + 1;
+                prevRank = rank;
+                leaderboardRow.find('#rank').html(rank);
+            }
+        }
+        // This give colour to rows where the student's rank is in the top 3       
+        leaderboardRow.attr('class', `rank-${rank <= 3 ? rank : 'default'}`);
+        leaderboardRow.find('#displayName').html(studentObject.displayName);
+        leaderboardRow.find('#criteria').html(studentObject[boardType.name] +
```
Comment (esteeroramy) [inline public/js/leaderboard.js]: this should be a separate block, for adding the`%` sign

**Comment by esteeroramy** [comment]
let's add the current user rank to the table, and highlight their name in a different colour. if they are past the limit that we can show then we can add a new row saying "..." and then we would have them on the row after. we should also give them a different colour than the standard white

Momo - Done, gave them the pink color like the nav bar. Let me know if you think a differnt colour would be better.

**Comment by esteeroramy** [comment]
Also, lets hook this up to general settings

Momo - Done

File: server/users.js [hunk 55]
```diff
@@ -430,89 +430,84 @@ exports.updateProfile = function (userId, request, callback) {
...
+                        (currentStudent.points/currentStudent.totalAttemptsCount))).toFixed(2)
                 }
+
+                leaderboardList.push(student);
             }
-            callback(lb);
-        } else {
-            callback(studentlist);
         }
+        callback(leaderboardList, displayRank);
```
Comment (ahmedqarmout2) [inline server/users.js]: just to be consistent please return an err, you can send null back

File: public/js/common.js (hunk 48 above)
Comment (Mohamed-A-Mohamed) [reply public/js/common.js]: Fixed

File: public/js/leaderboard.js (hunk 50 above)
Comment (Mohamed-A-Mohamed) [reply public/js/leaderboard.js]: Fixed

File: public/js/home.js (hunk 49 above)
Comment (Mohamed-A-Mohamed) [reply public/js/home.js]: Fixed

File: server.js (hunk 51 above)
Comment (Mohamed-A-Mohamed) [reply server.js]: My bad I goofed lol. I've removed them and added Pug to the name of my consts

File: server.js (hunk 52 above)
Comment (Mohamed-A-Mohamed) [reply server.js]: Fixed, leader was the list of students I was getting from the db. I've given it a better name, it is now leaderboardList.

File: public/js/leaderboard.js (hunk 54 above)
Comment (Mohamed-A-Mohamed) [reply public/js/leaderboard.js]: Fixed

File: server/users.js (hunk 55 above)
Comment (Mohamed-A-Mohamed) [reply server/users.js]: Fixed

File: public/js/leaderboard.js (hunk 53 above)
Comment (Mohamed-A-Mohamed) [reply public/js/leaderboard.js]: Fixed

**Review by ahmedqarmout2 — APPROVED** [review]

File: server/users.js [hunk 56]
```diff
@@ -429,89 +430,86 @@ exports.updateProfile = function (userId, request, callback) {
...
  * @param {function} callback
  */
-exports.getLeaderboard = function (userid, shrt, callback) {
+exports.getLeaderboard = function (userid, smallBoard, callback) {
     getStudentsListSorted(0, function(err, studentlist) {
         if (err) {
-            logger.error(err);
-            return (err, []);
+            logger.error('Leaderboard broke: ' + err);
+            callback(err, []);
```
Comment (Mohammed-JS-Khan) [inline server/users.js]: return callback(err, [])

File: server/users.js [hunk 57]
```diff
@@ -429,89 +430,86 @@ exports.updateProfile = function (userId, request, callback) {
...
+                        (currentStudent.points/currentStudent.totalAttemptsCount))).toFixed(2)
                 }
+
+                leaderboardList.push(student);
             }
-            callback(lb);
-        } else {
-            callback(studentlist);
         }
+        callback(err, leaderboardList);
```
Comment (Mohammed-JS-Khan) [inline server/users.js]: return callback... etc

**Review by Mohammed-JS-Khan — APPROVED** [review]

File: server/users.js (hunk 56 above)
Comment (Mohamed-A-Mohamed) [reply server/users.js]: Fixed

File: server/users.js (hunk 57 above)
Comment (Mohamed-A-Mohamed) [reply server/users.js]: Fixed

**Review by esteeroramy — APPROVED** [review]

**Pull request merged**